In [66]:
import pandas as pd
import numpy as np

In [67]:
autos = pd.read_csv('autos.csv', encoding = 'Latin=1')
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [68]:
print(autos.head(5))

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

From the work we did in the last screen, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [69]:
# print a array of the existing column names

autos_col = autos.columns
print(autos_col)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [70]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

print(autos.head(3))

          date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0   

In [71]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Price and odometer columns are numeric valures stored as text

In [72]:
print(autos["price"].head())
print(autos["odometer"].head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


In [73]:
autos['price'] = (autos['price']
                      .str.replace(',', '')
                      .str.replace('$','')
                      .astype(int)
                 )

autos.rename({'price':'usd_price'}, inplace=True, axis=1)
autos['usd_price'].head(5)


0    5000
1    8500
2    8990
3    4350
4    1350
Name: usd_price, dtype: int64

In [74]:
autos['odometer'] = (autos['odometer']
                         .str.replace('km', '')
                         .str.replace(',','')
                         .astype(int)
                    )

autos.rename({'odometer':'odometer_km'}, inplace=True, axis=1)
autos['odometer_km'].head(5)

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Describe outliers


In [75]:
autos[['odometer_km', 'usd_price']].describe()

odometer_km     usd_price
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08

### Observations
#### Odometer readings

- The median and 75th percentile value are the same as the maximum value (150,000km). The 25th percentile is set at around 125,000km.
- This indicates that a reading of 150,000km is not an outlier, since 75% of the listed cars are between 125,000 to 150,000km in mileage.
#### Price

- The maximum price is 100,000,000USD. This varies significantly from the median and 75th percentile values of 2,950 and 7,200USD respectively. The standard deviation is also considerably high (about 480,000 USD).
- The minimum price of 0 is unusual and could affect our analysis.

In [77]:
# odemeter_km
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

#### Additional Observations
- The least mileage observed is 5000km rather than 0km. This is expected because the cars listed on this platform are used cars.
- Higher Mileage cars (above 50,000km) are more common on the listing than lower mileage ones.
- Over 30,000 (~65%) cars have odometer readings of 150,000. This could imply that there is an upper mileage limit of 150,000km for cars accepted for listing.

In [79]:
# usd_price
autos['usd_price'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: usd_price, dtype: int64

In [80]:
autos['usd_price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: usd_price, dtype: float64

### Removing Price Outliers
As observed earlier, the 25th, 50th and 75th percentile of the prices are 1,100, 2,950 and 7,200USD. We will use a selection method to create a “data fence” from our 25th to 75th percentiles. Any data beyond this fence will be considered an outlier.

We already have enough information to calculate our interquartile range IQR. We will use the equation below to determine our upper and lower price fences:

    Interquartile Range (IQR) = 75th Percentile - 25th Percentile
    Lower fence = 25th Percentile - 1.5*(IQR)
    Upper fence = 75th Percentile + 1.5*(IQR)
Lets define all these boundaries in the code cell below:

In [81]:
price_25, price_75 = (1100, 7200)

IQR = price_75 - price_25
lower_fence = price_25 - (1.5*IQR)
upper_fence = price_75 + (1.5*IQR)

print('Interquartile Range: ', IQR)
print('Lower fence: ', lower_fence)
print('Upper fence: ', upper_fence)

Interquartile Range:  6100
Lower fence:  -8050.0
Upper fence:  16350.0


Fences may extend beyond the data sometimes, so it is totally fine if our lower fence is negative. We should be mindful, however, that a minimum price value of 0 can also affect our analysis. In this case, it is advisable to eliminate the entries with zero prices too.

Before we proceed, let's take a look at how many listings are recorded at zero price:

In [82]:
autos['usd_price'].value_counts().head()


0       1421
500      781
1500     734
2500     643
1000     639
Name: usd_price, dtype: int64

Of the 50,000 rows in our dataset, 1,421 records (3%) are listed at zero price. Since this percentage is very low, we can safely remove these entries. We will set our lower fence at 1 to exclude zero prices:

In [83]:
autos = autos[autos['usd_price'].between(1, upper_fence)]

autos['usd_price'].describe()

count    44795.000000
mean      4089.433620
std       3841.429247
min          1.000000
25%       1150.000000
50%       2700.000000
75%       5999.000000
max      16350.000000
Name: usd_price, dtype: float64

#### Observations
- Our dataset has been reduced to 44,795 rows in total. The 25th, 50th and 75th percentiles are close to their initial values before cleaning.
- The standard deviation (3,841USD) is considerably lower than the value of 480,000USD recorded before.
- 16,350USD is a better approximation for the maximum price than the previous value of 100,000,000USD, considering that our 75th percentile is around 6,000USD.

Let's now explore the date columns to understand the date range our data covers

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

```python 
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
```
> crawler = rastreador

In [84]:
# exploring the date columns
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [85]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025516
2016-03-06    0.013997
2016-03-07    0.036098
2016-03-08    0.033642
2016-03-09    0.033129
2016-03-10    0.032705
2016-03-11    0.032481
2016-03-12    0.037393
2016-03-13    0.015604
2016-03-14    0.036991
2016-03-15    0.034178
2016-03-16    0.029959
2016-03-17    0.031990
2016-03-18    0.012769
2016-03-19    0.034200
2016-03-20    0.037705
2016-03-21    0.037192
2016-03-22    0.032727
2016-03-23    0.032481
2016-03-24    0.029378
2016-03-25    0.032013
2016-03-26    0.032593
2016-03-27    0.030718
2016-03-28    0.034781
2016-03-29    0.034200
2016-03-30    0.033865
2016-03-31    0.031588
2016-04-01    0.032794
2016-04-02    0.034982
2016-04-03    0.038531
2016-04-04    0.036366
2016-04-05    0.012948
2016-04-06    0.003170
2016-04-07    0.001317
Name: date_crawled, dtype: float64

#### Observation:
Data is crawled between 5th March 2016 and 7th April 2016.


In [86]:
#Remove the values outside those upper and lower bounds and calculate the distribution of the remaining values using Series.value_counts(normalize=True).

autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
                ...   
2016-04-03    0.038799
2016-04-04    0.036701
2016-04-05    0.011653
2016-04-06    0.003237
2016-04-07    0.001183
Name: ad_created, Length: 73, dtype: float64

The dates are in between June 2015 and April 2016. We can analyze further based on month and year (without date)



In [87]:
autos["ad_created"].str[:7].value_counts(normalize=True, dropna=False).sort_index()


2015-08    0.000022
2015-09    0.000022
2015-11    0.000022
2015-12    0.000045
2016-01    0.000246
2016-02    0.001250
2016-03    0.839290
2016-04    0.159103
Name: ad_created, dtype: float64

#### Observation:
The maximum number of ads were created in the month of March 2016.

In [88]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()


2016-03-05    0.001161
2016-03-06    0.004576
2016-03-07    0.005693
2016-03-08    0.007903
2016-03-09    0.010068
2016-03-10    0.011162
2016-03-11    0.013037
2016-03-12    0.024936
2016-03-13    0.009376
2016-03-14    0.012859
2016-03-15    0.016274
2016-03-16    0.016966
2016-03-17    0.029043
2016-03-18    0.007568
2016-03-19    0.016475
2016-03-20    0.021230
2016-03-21    0.021185
2016-03-22    0.021922
2016-03-23    0.019109
2016-03-24    0.020449
2016-03-25    0.020002
2016-03-26    0.017234
2016-03-27    0.016207
2016-03-28    0.021699
2016-03-29    0.023016
2016-03-30    0.025472
2016-03-31    0.024333
2016-04-01    0.023306
2016-04-02    0.025226
2016-04-03    0.025673
2016-04-04    0.025047
2016-04-05    0.121286
2016-04-06    0.214332
2016-04-07    0.126175
Name: last_seen, dtype: float64

#### Observation:
On 5.April, 6.April and 7.April, we can see the spike in last seen entries,

### Phase 6: Analyzing Registration Year

In [90]:
autos["registration_year"].describe()

count    44795.00000
mean      2003.97339
std         74.81955
min       1000.00000
25%       1999.00000
50%       2003.00000
75%       2007.00000
max       9999.00000
Name: registration_year, dtype: float64

registration year 1000 and 9999 looks unrealistic. So let's find out the minimum logical registration year.

In [91]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index()


1000    0.000022
1001    0.000022
1111    0.000022
1800    0.000045
1910    0.000112
          ...   
4800    0.000022
5000    0.000089
5911    0.000022
8888    0.000022
9999    0.000045
Name: registration_year, Length: 85, dtype: float64

#### Observation:
1910 looks realistic minimum year and 2016 looks realistic maximum year because the registration date cannot be later than ad_created. So we can delete the entries having invalid data in registration year

In [92]:
autos = autos[autos["registration_year"].between(1910,2016)]

In [93]:
autos.describe()

usd_price  registration_year      power_ps    odometer_km  \
count  42959.000000       42959.000000  42959.000000   42959.000000   
mean    4129.176680        2002.395191    109.872925  129684.582974   
std     3863.748169           6.750207    188.566039   36385.094333   
min        1.000000        1910.000000      0.000000    5000.000000   
25%     1150.000000        1999.000000     69.000000  125000.000000   
50%     2750.000000        2003.000000    105.000000  150000.000000   
75%     6000.000000        2007.000000    140.000000  150000.000000   
max    16350.000000        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        42959.000000         42959.0  42959.000000  
mean             5.791918             0.0  50613.600549  
std              3.691957             0.0  25736.625285  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30165.000000  
50%              6.000000             0.0  49453.000000  
75%              9.000000             0.0  71144.000000  
max             12.000000             0.0  99998.000000

Around 1884 entries have been removed from the analysis.



## Exploring price by brand
We will use aggregation to understand the brand column.



In [94]:
autos["brand"].value_counts(normalize=True)


volkswagen        0.214763
opel              0.115203
bmw               0.103541
mercedes_benz     0.088131
audi              0.077562
ford              0.073093
renault           0.050769
peugeot           0.032170
fiat              0.027747
seat              0.019088
skoda             0.016993
mazda             0.016085
nissan            0.015782
smart             0.015363
citroen           0.015038
toyota            0.013362
hyundai           0.010545
volvo             0.009334
mitsubishi        0.008706
sonstige_autos    0.008496
honda             0.008334
mini              0.007752
alfa_romeo        0.007007
kia               0.006960
suzuki            0.006378
chevrolet         0.005750
chrysler          0.003748
dacia             0.002863
daihatsu          0.002724
subaru            0.002211
jeep              0.002002
porsche           0.001886
saab              0.001769
daewoo            0.001629
trabant           0.001513
rover             0.001420
land_rover        0.001304
j

In [96]:
#Let us consider the brands that form at least 1 % of our data.
brands_all = autos["brand"].value_counts(normalize=True)
brands_selected = brands_all[brands_all > 0.01].index

brand_price = {}
for brand in brands_selected:
    brand_autos = autos[autos["brand"]==brand]
    mean_price = brand_autos["usd_price"].mean()
    brand_price[brand] = int(mean_price)
    
brand_price

{'volkswagen': 4183,
 'opel': 2712,
 'bmw': 5649,
 'mercedes_benz': 5259,
 'audi': 5703,
 'ford': 2944,
 'renault': 2281,
 'peugeot': 2967,
 'fiat': 2740,
 'seat': 3747,
 'skoda': 5586,
 'mazda': 3557,
 'nissan': 3899,
 'smart': 3560,
 'citroen': 3543,
 'toyota': 4594,
 'hyundai': 4876}

#### Observations:
Audi, Mercedes Benz and BMW are the expensive ones but have maximum ads created also. This can be marked as high end cars.
Volkswagon is mid-level expensive and at the same time has the maximum entries. This is the most famous brand and cost efficient
Ford and Opel are least expensive with maximum ads created.

### Storing aggregate data in a data frame
Calculate mean mileage and mean price for the above selected brands which constitute more than 0.05% of our data and create the data frame. We will be caluclating these separately for automatic and manual.

In [98]:
brands_all = autos["brand"].value_counts(normalize=True)
brands_selected = brands_all[brands_all > 0.05].index

brand_price_automatik = {}
brand_price_manuell = {}
brand_odometer_automatik = {}
brand_odometer_manuell = {}
brand_count_manuell = {}
brand_count_automatik = {}
for brand in brands_selected:
    brand_autos = autos[(autos["brand"] == brand) & (autos["gearbox"] == "manuell")]
    brand_autos_count = brand_autos["brand"].value_counts()
    mean_price = brand_autos["usd_price"].mean()
    mean_mileage = brand_autos["odometer_km"].mean()
    brand_price_manuell[brand] = int(mean_price)
    brand_odometer_manuell[brand] = int(mean_mileage)
    brand_count_manuell[brand] = int(brand_autos_count)

    brand_autos = autos[(autos["brand"] == brand) & (autos["gearbox"] == "automatik")]
    brand_autos_count = brand_autos["brand"].value_counts()
    brand_autos_automatik_count = brand_autos["brand"].value_counts()
    mean_price = brand_autos["usd_price"].mean()
    mean_mileage = brand_autos["odometer_km"].mean()
    brand_price_automatik[brand] = int(mean_price)    
    brand_odometer_automatik[brand] = int(mean_mileage)
    brand_count_automatik[brand] = int(brand_autos_count)
    
bpa_series = pd.Series(brand_price_automatik)
bpm_series = pd.Series(brand_price_manuell)
boa_series = pd.Series(brand_odometer_automatik)
bom_series = pd.Series(brand_odometer_manuell)
bcm_series = pd.Series(brand_count_manuell)
bca_series = pd.Series(brand_count_automatik)

brand_summary = pd.DataFrame(bpa_series, columns=['mean_price_automatik'])
brand_summary["mean_odometer_automatik"] = boa_series
brand_summary["count_automatik"] = bca_series
brand_summary["mean_price_manuell"] = bpm_series
brand_summary["mean_odometer_manuell"] = bom_series
brand_summary["count_manuell"] = bcm_series

brand_summary

mean_price_automatik  mean_odometer_automatik  count_automatik  \
volkswagen                     6620                   131800             1011   
opel                           2966                   128069              492   
bmw                            7288                   141112             1254   
mercedes_benz                  6323                   139706             1890   
audi                           7201                   141227              925   
ford                           4139                   128233              201   
renault                        3251                   122769              139   

               mean_price_manuell  mean_odometer_manuell  count_manuell  
volkswagen                   3971                 133139           7736  
opel                         2747                 131007           4215  
bmw                          5074                 138577           3055  
mercedes_benz                4260                 138334           1762  
audi                         5269                 139488           2260  
ford                         2912                 127170           2788  
renault                      2274                 129249           1933

#### Observations:
The mean price for automatik cars is almost double that of manuell in any given Brand except for Opel, where the count of manuell cars is almost eight times of automatik cars
The count of manuell cars listed are more than count of automatik cars listed except for Merceds_Benz. Also count of Mercedes_Benz Automatik is the highest in Automatik.
There is not much variation in odometer based on brand or gearbox